### PR TITLE
Tpietruszka feature/support optimizers requiring closure arguments

### PR DESCRIPTION
**Pull request by benjamin-work**
I didn't manage to push on the existing PR, so here is a new one -- sorry for that.

### CODE REVIEW

File: skorch/tests/test_net.py [hunk 1]
```diff
@@ -1409,6 +1419,70 @@ def test_batch_size_neg_1_uses_whole_dataset(
...
+            """Call NLLLoss 3 times."""
+            loss = torch.nn.NLLLoss()
+            def __call__(self, *args, **kwargs):
+                loss = self.loss(*args, **kwargs)
+                side_effect.append(float(loss))
+                loss = self.loss(*args, **kwargs)
+                side_effect.append(float(loss))
+                loss = self.loss(*args, **kwargs)
+                side_effect.append(float(loss))
+                return loss
```
Comment (ottonemo) [inline skorch/tests/test_net.py]: This should be a custom optimizer instead of a criterion as the additional loss values do not contribute to the test.

**Comment by benjamin-work** [comment]
@ottonemo pls review again

**Comment by benjamin-work** [comment]
great work @tpietruszka @ottonemo I believe the long discussion was worth it at the end.
